src/cascade/model/model.py

```diff
@@ -21,31 +25,52 @@ def __init__(self, nonzero_rates, parent_location, child_location=None, covariat
         self.nonzero_rates = nonzero_rates
         self.location_id = parent_location
         self.child_location = child_location if child_location else list()
+        double_locations = [l for (l, v) in Counter(self.child_location + [parent_location]).items() if v > 1]
+        if double_locations:
+            raise ValueError(f"Multiple locations have same ID {double_locations}")
         # Covariates are here because their reference values are part of
         # the model. Even though avgint and data use them, a model is always
         # written before the avgint and data are written.
         self.covariates = covariates if covariates else list()
+        self._check_covariates(self.covariates)
         # There are always four weights, constant, susceptible,
         # with_condition, and total.
         if weights:
             self.weights = weights
         else:
             self.weights = dict()
+        self._check_weights(self.weights)
         self._scale = None
         self.scale_set_by_user = False
 
     @property
     def scale(self):
+        """The scale is a Var, so it has a value for every model variable.
+        It is the value of the model variable at which to evaluate the
+        derivative of its base log-likelihood. This derivative sets the
+        baseline against which nonlinear optimization will compare later
+        values. If the derivative is zero, then the optimization will ignore
+        this variable.
+        """
         return self._scale
 
     @scale.setter
     def scale(self, value):
-        # Dismod-AT usually calculates this for you. If you've set it by
-        # hand, this records that this is the case so that it will rewrite
-        # what Dismod-AT calculates.
+        """Dismod-AT usually calculates this for you. If you've set it by
+        hand, this records that this is the case so that it will rewrite
+        what Dismod-AT calculates."""
         self.scale_set_by_user = True
         self._scale = value
 
+    def model_like(self):
```

Comment:
It's the usage of the object that's confusing, because the author is confused. Can SmoothGrids be shared among models, in which case changing one SmoothGrid changes the grid under many models? Or do we enforce that they are separate? I feel the Pythonic way is to make copy() a shallow copy() and then it's on the user if they don't realize it was shallow.

Reply:
But is shallow copy ever what you want? I don't think it is. So it should be a deep copy. Does it make sense to use the copy constructor syntax like dict and list do? It makes your constructor more complex because you're effectively overloading it. Howabout a static method: `Model.from_model(other_model)`?